File: Medium/longest_subtstring_without_repeating_characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function (s) {
  if (s == "" || s.length > 50000) {
    return 0;
  }

  let longestSubtstring = 1;
  let startIndex = {};
  let startPosition = 0;
  const strSplit = s.split("");
 
  strSplit.forEach((c, i) => {
    if(Object.hasOwn(startIndex, c)) {
        if(startIndex[c] >= startPosition) {
            startPosition = startIndex[c] + 1;
        }
    }

    startIndex[c] = i;
    longestSubtstring = Math.max(longestSubtstring, i - startPosition + 1);
  });

  return longestSubtstring;
};

console.log(lengthOfLongestSubstring("abcabcbb")); //3
console.log(lengthOfLongestSubstring("bbbbb")); //1
console.log(lengthOfLongestSubstring("pwwkew")); //3
console.log(lengthOfLongestSubstring("")); //0
console.log(lengthOfLongestSubstring("au")); //2
console.log(lengthOfLongestSubstring("aab")); //2

File: Medium/longest_subtstring_without_repeating_characters.go
package main

import (
	"fmt"
	"strings"
)

func lengthOfLongestSubstring(s string) int {
	if len(s) == 0 || len(s) > 50000 {
		return 0
	}

	startIndex := make(map[string]int)
	longestSubtstring := 1
	startPosition := 0
	strSplit := strings.Split(s, "")

	for i, c := range strSplit {
		if _, ok := startIndex[c]; ok {
			if startIndex[c] >= startPosition {
				startPosition = startIndex[c] + 1
			}
		}

		startIndex[c] = i
		longestSubtstring = max(longestSubtstring, i-startPosition+1)
	}

	return longestSubtstring
}

func run_lengthOfLongestSubstring() {
	fmt.Println(lengthOfLongestSubstring("abcabcbb")) // 3
	fmt.Println(lengthOfLongestSubstring("bbbbb"))    // 1
	fmt.Println(lengthOfLongestSubstring("pwwkew"))   // 3
	fmt.Println(lengthOfLongestSubstring(""))         // 0
	fmt.Println(lengthOfLongestSubstring("au"))       // 2
	fmt.Println(lengthOfLongestSubstring("aab"))      // 2
}
